File: lp4k-emu-web/src/test/java/net/thecodersbreakfast/lp4j/emulator/EmulatorLaunchpadClientTest.java
/*
 *    Copyright 2020 Vasyl Rudas
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package net.thecodersbreakfast.lp4j.emulator;

import io.lp4k.emulator.MockEventBus;
import io.lp4k.emulator.MockVertx;
import io.vertx.core.json.JsonObject;
import net.thecodersbreakfast.lp4j.api.*;
import io.lp4k.emulator.output.OutputEventType;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

import static java.util.Collections.emptyMap;
import static java.util.Map.entry;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.*;

class EmulatorLaunchpadClientTest {

    private static final String BRIGHTNESS_KEY = "b";
    private static final String LIGHT_INTENSITY_KEY = "i";
    private static final String OPERATION_KEY = "o";

    private static final String COLOR_KEY = "c";
    private static final String RED_COLOR_KEY = "r";
    private static final String GREEN_COLOR_KEY = "g";

    private static final String PAD_X_KEY = "x";
    private static final String PAD_Y_KEY = "y";

    private static final String BUTTON_COORDINATE_KEY = "i";
    private static final String BUTTON_IS_TOP_KEY = "t";

    private static final String VISIBLE_BUFFER_KEY = "v";
    private static final String WRITE_BUFFER_KEY = "w";
    private static final String COPY_VISIBLE_BUFFER_KEY = "c";
    private static final String AUTO_SWAP_BUFFERS_KEY = "a";

    void verifyEventWasPublished(
        OutputEventType outputEventType,
        Map<String, Object> expectedParametersForPublish
    ) {
        Map<String, Object> parametersToCheck = new HashMap<>();

        parametersToCheck.put("evt", outputEventType.name());
        parametersToCheck.putAll(expectedParametersForPublish);

        var message = new JsonObject(parametersToCheck);
        verify(eventBus).publish("lp4j:client", message);
    }

    private MockEventBus eventBus;

    private EmulatorLaunchpadClient emulatorLaunchpadClient;

    @BeforeEach
    void setUp() {
        MockVertx vertx = mock(MockVertx.class);

        eventBus = mock(MockEventBus.class);
        when(vertx.eventBus()).thenReturn(eventBus);

        emulatorLaunchpadClient = new EmulatorLaunchpadClient(vertx);
    }

    @Test
    void reset_event_was_published() {
        emulatorLaunchpadClient.reset();

        verifyEventWasPublished(OutputEventType.RST, emptyMap());
    }

    @Test
    void set_lights_event_not_published_because_not_implemented() {
        assertThrows(
            UnsupportedOperationException.class,
            () -> emulatorLaunchpadClient.setLights(new Color[0], BackBufferOperation.NONE)
        );
    }

    @Test
    void test_lights_event_not_published_because_of_null_light_intensity() {
        assertThrows(
            IllegalArgumentException.class,
            () -> emulatorLaunchpadClient.testLights(null)
        );
    }

    @Test
    void test_lights_event_published_with_low_light_intensity() {
        emulatorLaunchpadClient.testLights(LightIntensity.LOW);

        verifyEventWasPublished(
            OutputEventType.TST,
            Map.of(LIGHT_INTENSITY_KEY, 5)
        );
    }

    @Test
    void test_lights_event_published_with_medium_light_intensity() {
        emulatorLaunchpadClient.testLights(LightIntensity.MEDIUM);

        verifyEventWasPublished(
            OutputEventType.TST,
            Map.of(LIGHT_INTENSITY_KEY, 10)
        );
    }

    @Test
    void test_lights_event_published_with_high_light_intensity() {
        emulatorLaunchpadClient.testLights(LightIntensity.HIGH);

        verifyEventWasPublished(
            OutputEventType.TST,
            Map.of(LIGHT_INTENSITY_KEY, 15)
        );
    }

    @Test
    void set_pad_light_event_not_published_because_of_null_pad_argument() {
        assertThrows(
            IllegalArgumentException.class,
            () -> emulatorLaunchpadClient.setPadLight(null, null, null)
        );
    }

    @Test
    void set_pad_light_event_not_published_because_of_null_color_argument() {
        assertThrows(
            IllegalArgumentException.class,
            () -> emulatorLaunchpadClient.setPadLight(Pad.at(0, 0), null, null)
        );
    }

    @Test
    void set_pad_light_event_not_published_because_of_null_operation_argument() {
        assertThrows(
            IllegalArgumentException.class,
            () -> emulatorLaunchpadClient.setPadLight(
                Pad.at(0, 0),
                Color.RED,
                null
            )
        );
    }

    @Test
    void set_pad_light_event_published() {
        emulatorLaunchpadClient.setPadLight(Pad.at(0, 0), Color.RED, BackBufferOperation.NONE);

        verifyEventWasPublished(
            OutputEventType.PADLGT,
            Map.ofEntries(
                entry(PAD_X_KEY, 0),
                entry(PAD_Y_KEY, 0),
                entry(
                    COLOR_KEY,
                    new JsonObject()
                        .put(RED_COLOR_KEY, 3)
                        .put(GREEN_COLOR_KEY, 0)
                ),
                entry(OPERATION_KEY, "NONE")
            )
        );
    }

    @Test
    void set_button_light_event_not_published_because_of_null_pad_argument() {
        assertThrows(
            IllegalArgumentException.class,
            () -> emulatorLaunchpadClient.setButtonLight(null, null, null)
        );
    }

    @Test
    void set_button_light_event_not_published_because_of_null_color_argument() {
        assertThrows(
            IllegalArgumentException.class,
            () -> emulatorLaunchpadClient.setButtonLight(Button.UP, null, null)
        );
    }

    @Test
    void set_button_light_event_not_published_because_of_null_operation_argument() {
        assertThrows(
            IllegalArgumentException.class,
            () -> emulatorLaunchpadClient.setButtonLight(
                Button.VOL,
                Color.GREEN,
                null
            )
        );
    }

    @Test
    void set_button_light_event_published_for_right_button() {
        emulatorLaunchpadClient.setButtonLight(Button.VOL, Color.GREEN, BackBufferOperation.CLEAR);

        verifyEventWasPublished(
            OutputEventType.BTNLGT,
            Map.ofEntries(
                entry(BUTTON_COORDINATE_KEY, 0),
                entry(BUTTON_IS_TOP_KEY, false),
                entry(
                    COLOR_KEY,
                    new JsonObject()
                        .put(RED_COLOR_KEY, 0)
                        .put(GREEN_COLOR_KEY, 3)
                ),
                entry(OPERATION_KEY, "CLEAR")
            )
        );
    }

    @Test
    void set_button_light_event_published_for_top_button() {
        emulatorLaunchpadClient.setButtonLight(Button.UP, Color.GREEN, BackBufferOperation.COPY);

        verifyEventWasPublished(
            OutputEventType.BTNLGT,
            Map.ofEntries(
                entry(BUTTON_COORDINATE_KEY, 0),
                entry(BUTTON_IS_TOP_KEY, true),
                entry(
                    COLOR_KEY,
                    new JsonObject()
                        .put(RED_COLOR_KEY, 0)
                        .put(GREEN_COLOR_KEY, 3)
                ),
                entry(OPERATION_KEY, "COPY")
            )
        );
    }

    @Test
    void set_brightness_event_not_published_because_of_null_brightness_argument() {
        assertThrows(
            IllegalArgumentException.class,
            () -> emulatorLaunchpadClient.setBrightness(null)
        );
    }

    @Test
    void set_brightness_event_published() {
        emulatorLaunchpadClient.setBrightness(Brightness.BRIGHTNESS_MAX);

        verifyEventWasPublished(
            OutputEventType.BRGHT,
            Map.of(
                BRIGHTNESS_KEY,
                Brightness.BRIGHTNESS_MAX.getBrightnessLevel()
            )
        );
    }

    @Test
    void set_buffers_event_not_published_because_of_null_visible_buffer_argument() {
        assertThrows(
            IllegalArgumentException.class,
            () -> emulatorLaunchpadClient.setBuffers(
                null,
                null,
                false,
                false
            )
        );
    }

    @Test
    void set_buffers_event_not_published_because_of_null_write_buffer_argument() {
        assertThrows(
            IllegalArgumentException.class,
            () -> emulatorLaunchpadClient.setBuffers(
                Buffer.BUFFER_0,
                null,
                false,
                false
            )
        );
    }

    @Test
    void set_buffers_event_published() {
        emulatorLaunchpadClient.setBuffers(
            Buffer.BUFFER_1,
            Buffer.BUFFER_0,
            true,
            true
        );

        verifyEventWasPublished(
            OutputEventType.BUF,
            Map.ofEntries(
                entry(VISIBLE_BUFFER_KEY, Buffer.BUFFER_1.name()),
                entry(WRITE_BUFFER_KEY, Buffer.BUFFER_0.name()),
                entry(COPY_VISIBLE_BUFFER_KEY, true),
                entry(AUTO_SWAP_BUFFERS_KEY, true)
            )
        );
    }

    @Test
    void scroll_text_event_not_published_because_not_implemented() {
        assertThrows(
            UnsupportedOperationException.class,
            () -> emulatorLaunchpadClient.scrollText(
                "",
                Color.AMBER,
                ScrollSpeed.SPEED_MIN,
                false,
                BackBufferOperation.NONE
            )
        );
    }
}

File: lp4k-emu-web/src/test/kotlin/io/lp4k/emulator/EmulatorLaunchpadClientTest.kt
/*
 *    Copyright 2020 Vasyl Rudas
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */
package io.lp4k.emulator

import io.lp4k.emulator.output.OutputEventType
import io.lp4k.launchpad.api.BackBufferOperation
import io.lp4k.launchpad.api.Buffer
import io.lp4k.launchpad.api.s.*
import io.vertx.core.json.JsonObject
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.mockito.Mockito.*

@Suppress("kotlin:S100")
internal class EmulatorLaunchpadClientTest {

    private fun verifyEventWasPublished(
        outputEventType: OutputEventType,
        expectedParametersForPublish: Map<String, Any?>
    ) {
        val parametersToCheck = mutableMapOf<String, Any?>(
            "evt" to outputEventType.name
        ) + expectedParametersForPublish

        val message = JsonObject(parametersToCheck)

        verify(eventBus).publish("lp4j:client", message)
    }

    private lateinit var eventBus: MockEventBus
    private lateinit var emulatorLaunchpadClient: EmulatorLaunchpadClient

    @BeforeEach
    fun setUp() {
        val vertx = mock(MockVertx::class.java)
        eventBus = mock(MockEventBus::class.java)

        `when`(vertx.eventBus()).thenReturn(eventBus)

        emulatorLaunchpadClient = EmulatorLaunchpadClient(vertx)
    }

    @Test
    fun `reset event was published`() {
        emulatorLaunchpadClient.reset()

        verifyEventWasPublished(OutputEventType.RST, emptyMap())
    }

    @Test
    fun `set lights event not published because not implemented`() {
        assertThrows<UnsupportedOperationException> {
            emulatorLaunchpadClient.setLights(
                emptyArray(),
                BackBufferOperation.NONE
            )
        }
    }

    @Test
    fun `test lights event published with low light intensity`() {
        emulatorLaunchpadClient.testLights(LightIntensityLaunchS.LOW)

        verifyEventWasPublished(
            OutputEventType.TST,
            java.util.Map.of<String, Any>(LIGHT_INTENSITY_KEY, 5)
        )
    }

    @Test
    fun `test lights event published with medium light intensity`() {
        emulatorLaunchpadClient.testLights(LightIntensityLaunchS.MEDIUM)

        verifyEventWasPublished(
            OutputEventType.TST,
            mapOf(LIGHT_INTENSITY_KEY to 10)
        )
    }

    @Test
    fun `test lights event published with high light intensity`() {
        emulatorLaunchpadClient.testLights(LightIntensityLaunchS.HIGH)

        verifyEventWasPublished(
            OutputEventType.TST,
            mapOf(LIGHT_INTENSITY_KEY to 15)
        )
    }

    @Test
    fun `set pad light event published`() {
        emulatorLaunchpadClient.setPadLight(
            PadLaunchS.at(0, 0),
            ColorLaunchS.RED,
            BackBufferOperation.NONE
        )

        verifyEventWasPublished(
            OutputEventType.PADLGT,
            mapOf(
                PAD_X_KEY to 0,
                PAD_Y_KEY to 0,
                COLOR_KEY to JsonObject()
                    .put(RED_COLOR_KEY, 3)
                    .put(GREEN_COLOR_KEY, 0),
                OPERATION_KEY to "NONE"
            )
        )
    }

    @Test
    fun `set button light event published for right button`() {
        emulatorLaunchpadClient.setButtonLight(
            ButtonLaunchS.VOL,
            ColorLaunchS.GREEN,
            BackBufferOperation.CLEAR
        )

        verifyEventWasPublished(
            OutputEventType.BTNLGT,
            mapOf(
                BUTTON_COORDINATE_KEY to 0,
                BUTTON_IS_TOP_KEY to false,
                COLOR_KEY to
                    JsonObject()
                        .put(RED_COLOR_KEY, 0)
                        .put(GREEN_COLOR_KEY, 3),
                OPERATION_KEY to "CLEAR"
            )
        )
    }

    @Test
    fun `set button light event published for top button`() {
        emulatorLaunchpadClient.setButtonLight(
            ButtonLaunchS.UP,
            ColorLaunchS.GREEN,
            BackBufferOperation.COPY
        )

        verifyEventWasPublished(
            OutputEventType.BTNLGT,
            mapOf(
                BUTTON_COORDINATE_KEY to 0,
                BUTTON_IS_TOP_KEY to true,
                COLOR_KEY to JsonObject()
                    .put(RED_COLOR_KEY, 0)
                    .put(GREEN_COLOR_KEY, 3)
                ,
                OPERATION_KEY to "COPY"
            )
        )
    }

    @Test
    fun `set brightness event published`() {
        emulatorLaunchpadClient.setBrightness(BrightnessLaunchS.BRIGHTNESS_MAX)

        verifyEventWasPublished(
            OutputEventType.BRGHT,
            mapOf(BRIGHTNESS_KEY to BrightnessLaunchS.BRIGHTNESS_MAX.brightnessLevel)
        )
    }

    @Test
    fun `set buffers event published`() {
        emulatorLaunchpadClient.setBuffers(
            Buffer.BUFFER_1,
            Buffer.BUFFER_0,
            copyVisibleBufferToWriteBuffer = true,
            autoSwap = true
        )

        verifyEventWasPublished(
            OutputEventType.BUF,
            mapOf(
                VISIBLE_BUFFER_KEY to Buffer.BUFFER_1.name,
                WRITE_BUFFER_KEY to Buffer.BUFFER_0.name,
                COPY_VISIBLE_BUFFER_KEY to true,
                AUTO_SWAP_BUFFERS_KEY to true
            )
        )
    }

    @Test
    fun `scroll text event not published because not implemented`() {
        assertThrows<UnsupportedOperationException> {
            emulatorLaunchpadClient.scrollText(
                "",
                ColorLaunchS.AMBER,
                ScrollSpeedLaunchS.SPEED_MIN,
                false,
                BackBufferOperation.NONE
            )
        }
    }

    companion object {
        private const val BRIGHTNESS_KEY = "b"

        private const val LIGHT_INTENSITY_KEY = "i"

        private const val OPERATION_KEY = "o"

        private const val COLOR_KEY = "c"
        private const val RED_COLOR_KEY = "r"
        private const val GREEN_COLOR_KEY = "g"

        private const val PAD_X_KEY = "x"
        private const val PAD_Y_KEY = "y"

        private const val BUTTON_COORDINATE_KEY = "i"
        private const val BUTTON_IS_TOP_KEY = "t"

        private const val VISIBLE_BUFFER_KEY = "v"
        private const val WRITE_BUFFER_KEY = "w"
        private const val COPY_VISIBLE_BUFFER_KEY = "c"
        private const val AUTO_SWAP_BUFFERS_KEY = "a"
    }
}